overlord/snapstate/snapmgr.go

```diff
@@ -502,6 +502,90 @@ func (m *SnapManager) ensureForceDevmodeDropsDevmodeFromState() error {
 	return nil
 }
 
+// ensureSnapdSnapTransition will migrate systems to use the "snapd" snap
+func (m *SnapManager) ensureSnapdSnapTransition() error {
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	// check if snapd snap is installed
+	var snapst SnapState
+	err := Get(m.state, "snapd", &snapst)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+	// nothing to do
+	if snapst.IsInstalled() {
+		return nil
+	}
+
+	// check if the user opts into the snapd snap
+	experimentalAllowSnapd, err := optedIntoSnapdSnap(m.state)
+	if err != nil {
+		return err
+	}
+	// nothing to do: the user does not want the snapd snap yet
+	if !experimentalAllowSnapd {
+		return nil
+	}
+
+	// ensure we only transition systems that have snaps already
+	installedSnaps, err := All(m.state)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+	// no installed snaps (yet): do nothing (fresh classic install)
+	if len(installedSnaps) == 0 {
+		return nil
+	}
+
+	// get current core snap and use same channel/user for the snapd snap
+	err = Get(m.state, "core", &snapst)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+	coreChannel := snapst.Channel
+	userID := snapst.UserID
+
+	// check that there is no change in flight already, this is a
+	// precaution to ensure the snapd transition is safe
+	for _, chg := range m.state.Changes() {
```

Comment:
we have this kind of code in many places now, wondering if it is time to introduce a helper